aspnetcore/web-api/action-return-types.md

```diff
@@ -10,6 +10,206 @@ uid: web-api/action-return-types
 ---
 # Controller action return types in ASP.NET Core web API
 
+:::moniker range=">= aspnetcore-7.0"
+
+[View or download sample code](https://github.com/dotnet/AspNetCore.Docs.Samples/tree/main/mvc/action-return-types/7.x/WebApiSample) ([how to download](xref:index#how-to-download-a-sample))
+
+ASP.NET Core provides the following options for web API controller action return types:
+
+* [Specific type](#specific-type)
+* [IActionResult](#iactionresult-type)
+* [ActionResult\<T>](#actionresultt-type)
+* [HttpResults](#httpresults-type)
+
+This article explains when it's most appropriate to use each return type.
+
+## Specific type
+
+The most basic action returns a primitive or complex data type, for example, `string` or a custom object. Consider the following action, which returns a collection of custom `Product` objects:
+
+:::code language="csharp" source="~/../AspNetCore.Docs.Samples/mvc/action-return-types/7.x/WebApiSample/Controllers/ProductsController.cs" id="snippet_Get":::
+
+Without known conditions to safeguard against, returning a specific type could suffice. The preceding action accepts no parameters, so parameter constraints validation isn't needed.
+
+When multiple return types are possible, it's common to mix an <xref:Microsoft.AspNetCore.Mvc.ActionResult> return type with the primitive or complex return type. Either [IActionResult](#iactionresult-type) or [ActionResult\<T>](#actionresultt-type) are necessary to accommodate this type of action. Several samples of multiple return types are provided in this article.
+
+### Return IEnumerable\<T> or IAsyncEnumerable\<T>
+
+ASP.NET Core buffers the result of actions that return <xref:System.Collections.Generic.IEnumerable%601> before writing them to the response. Consider declaring the action signature's return type as <xref:System.Collections.Generic.IAsyncEnumerable%601> to guarantee asynchronous iteration. Ultimately, the iteration mode is based on the underlying concrete type being returned. MVC automatically buffers any concrete type that implements `IAsyncEnumerable<T>`.
+
+Consider the following action, which returns sale-priced product records as `IEnumerable<Product>`:
+
+:::code language="csharp" source="~/../AspNetCore.Docs.Samples/mvc/action-return-types/7.x/WebApiSample/Controllers/ProductsController.cs" id="snippet_GetOnSaleProducts":::
+
+The `IAsyncEnumerable<Product>` equivalent of the preceding action is:
+
+:::code language="csharp" source="~/../AspNetCore.Docs.Samples/mvc/action-return-types/7.x/WebApiSample/Controllers/ProductsController.cs" id="snippet_GetOnSaleProductsAsync":::
+
```

Comment:
This section doesn't provide any guidance on when to use async. It asks the question and doesn't answer it.
> ASP.NET Core buffers the result of actions that return `IEnumerable<T>`
> MVC automatically buffers any concrete type that implements `IAsyncEnumerable<T>.`

So they're both buffered.
Do we want to say anything about buffering vs streaming?  In general, streaming is preferred.
There's also `IEnumerable<Task<T>>`

Reply:
Actually, it does not buffer anymore, since .NET 6 (https://learn.microsoft.com/en-us/dotnet/core/compatibility/aspnet-core/6.0/iasyncenumerable-not-buffered-by-mvc). Let me update the text.